### ISSUE TITLE
Filters on lists

### ISSUE THREAD

**Issue by Alexgmin** [Laravel-Backpack/CRUD#175]
This type of filters
![Image](https://cloud.githubusercontent.com/assets/8340441/19266741/5118e920-8fab-11e6-8c62-92f1386e6053.png)
(This screenshot is from https://github.com/zofe/rapyd-laravel)

The code to do this on rapyd

``` php
$filter->add('name', 'Name', 'text');

$filter->add('featured','Featured', 'select')
    ->options([null => 'Enabled', 1 => 'Enabled: Yes', 0 => 'Enabled: No']);

$filter->add('state','State', 'select')->options(
    Item::getStates()
);
$filter->add('created_at','Created', 'daterange')
    ->format('m/d/Y', 'en');

$filter->add('brand.name', 'Brand', 'select')
    ->options(
        Brands::getBrands()
    );

$filter->add('user.id', 'User ID', 'text')->clause('where')->operator('=');
```

What would be the best way to implement something like this on backpack?
Are there plans to do something similar?

**Comment by OwenMelbz** [comment]
This is a feature that's in development/planning

However if you add your own UI code into the template, you can utilise the "addClause()" method described here -> https://laravel-backpack.readme.io/docs/crud-full-api#usage-level-2-often

an example could be -> https://github.com/Laravel-Backpack/CRUD/issues/83

**Comment by tabacitu** [comment]
Hi @Alexgmin ,

There's no one-line way to do it right now. It's a know feature request and one of the 2 core features that are urgent (multi-language CRUDs, filters). I like how it's done in rapyd, thank you, we should take a closer look before we implement it ourselves and learn from it. 

For now, you can "fake" the filter functionality by adding new buttons on top of the form (next to the Add button). You can do that by adding buttons that, instead of being actual buttons, will be dropdowns or inputs.

So something like this should work:
```php

public function setup()
{
  // add a button whose HTML is in a view placed at resources\views\vendor\backpack\crud\buttons
  $this->crud->addButtonFromView('top', 'featured_filter', 'featured_filter', 'end');

  if (Input::get('featured_filter')) {
    $this->crud->addClause(Input::get('featured_filter'));
  }
}
```

resources/views/vendor/backpack/crud/buttons/featured_filter.blade.php:

```html
<select name="featured_filter">
  <option value="featured">Featured</option>
  <option value="not_featured">Not Featured</option>
  <option value="">Whatever</option>
</select>

<script>
jQuery(document).ready(function($) {
		$("select[name=featured_filter]").change(function() {
                     window.location.href = "{{ url($crud->route) }}?featured_filter="+$(this).val();
                 });
	});
</script>
```

Please note I haven't tested the code above, it's more like pseudocode :-)

**Comment by Alexgmin** [comment]
Yeah, that should work for now. Thanks, @tabacitu . Backpack is awesome, btw, we love it.

**Comment by tabacitu** [comment]
I think we could implement something very similar to how fields, buttons and columns work: filter types, each in their own blade file in ```/src/resources/views/filters/```. That would help in terms of familiarity and extensibility.

```php
$this->crud->addFilter([
  'name' => 'featured',
  'type' => 'dropdown',
  'values' => [
               'featured' => 'Featured',
               'not_featured' => 'Not featured',
               '' => 'Whatever',
              ],
], function($value) {
  if ($value != '') {
    $this->crud->addClause($value);
  }
});
```

The filters would come with some pre-built logic (probably) and you can pass a closure as an optional second parameter to overwrite it with your own.

---

What WOULD be very useful, and I hope it's ok if I hijack this thread for it, would be to make a list of all possible filter types. So that we know if this implementation is ok. //TODO

**Comment by tabacitu** [comment]
Also, I think we should allow some flexibility to _where_ those filters show up. Just like you can add buttons in different stacks, we might create special stacks for filters (or create more stacks for both buttons and filters).

I think useful places where we could add filters would be:
![screen_shot_2016-10-25_at_09_52_04](https://cloud.githubusercontent.com/assets/1032474/19675845/6c9ea254-9a99-11e6-80c1-d82a4bb76055.png)

Thoughts?

**Comment by mpixelz** [comment]
just my 2cents.. what about having a separate new row at the top after the title and before the actual table row?
something like this:
![screenshot-filters](https://cloud.githubusercontent.com/assets/22141147/19957016/56c6a964-a1b4-11e6-93e7-14671502d583.png)

**Comment by tabacitu** [comment]
@mpixelz yep, that could be an option too. So 4 stacks should do:
- header
- top_right
- center
- left
- right

**Comment by OwenMelbz** [comment]
@tabacitu I vaguely remember you saying you're working on this? is that still the case?

**Comment by tabacitu** [comment]
Yup. WIll be done by Nov 17th, the latest. Currently my thinking is to have these field types:

**Comment by tabacitu** [comment]
Currently my thinking is to first develop these most important filter types:
- select
- select2
- date
- daterange
- formula

The general syntax would be:
```php
$this->crud->addFilter($options, $values, $filter_logic);
```

I'll be releasing the select and select2 in about two weeks, and their syntax should be something like this:
```php
$this->crud->addFilter([ // select from enum column values
          'name' => 'published',
          'type' => 'select',
          'label'=> 'Published'
        ],
        function() {
            return App\Models\Article::getPossibleEnumValues('published');
        }
        // , function($value) {} // optional filter logic overwrite; by default it makes sure that {name} == $value
        );

$this->crud->addFilter([ // select_from_array
          'name' => 'featured',
          'type' => 'select_from_array',
        ], [ // values
           'featured' => 'Featured',
           'not_featured' => 'Not featured',
           '' => 'Whatever',
          ]
        // , function($value) {} // optional filter logic overwrite; by default it makes sure that {name} == $value
        });

$this->crud->addFilter([ // select_from_array
          'name' => 'featured',
          'type' => 'select_from_array',
        ], function() {
            // get all Users and create an associative array with ['id' => 'name']
            return App\Models\User::all()->keyBy('id')->pluck('name');
        }
        // , function($value) {} // optional filter logic overwrite; by default it makes sure that {name} == $value
        });
```

This would eliminate the need for a bunch of filter types for different purposes. What we'll do instead is give a number of examples, to match most common uses:
- enum
- select_from_array
- select_from_model
- select_from_ajax

What do you guys think?

**Comment by OwenMelbz** [comment]
Not sure if this is simplifying it too much, but could we utilise [model scopes](https://laravel.com/docs/5.3/eloquent#query-scopes)? for the actual filtering? maybe just define a scope, then internally handle the rest?

what would the select from ajax do? would it not be better to just do that by default?

additionally I've found https://github.com/dangrossman/bootstrap-daterangepicker to be good plugin for date ranges, specifically the "last 30 days" etc type fields.

but apart from that all happy here :)

**Comment by tabacitu** [comment]
Yup, you could move the filtering logic entirely into scopes with something like this:

```php
$this->crud->addFilter([ // select from enum column values
          'name' => 'published',
          'type' => 'select',
          'label'=> 'Published'
        ],
        function() {
            return App\Models\Article::getPossibleEnumValues('published');
        },
        function($value) {
            if ($value != '') {
              $this->crud->addClause($value); // adds the name of the variable as a scope
            }
        }
        );
```

Oooor... we could use the third parameter as either a closure or string. If you pass a closure, it will run the closure. If you pass a string, it will run a pre-defined operation. We could use these strings (similar to the PHP comparison operators):
- "==" -  db_value == selected_value [this would be the default]
- ">" - db_value > selected_value
- ">="
- "<"
- "<="
- "<>"
- "!="
- "like" -> db_value like %selected_value%
- "scope" - add the value as a scope
- anything else?

Not 100% sure this is achievable though.

Definitely agree to dangrossman's daterangepicker. It's awesome.

The select_from_ajax would load the select2 values with ajax. For when you have 1mil users and you want to filter the ones with the first_name==Owen. We can't do them all like this, because it would be too inconvenient to insert a filter then. In order to have an ajax filter you need to:
- add the filter (similar syntax, but you also give out the route to the ajax results)
- add a route (ex: /product/ajax-names)
- create a method ( ex: ProductCrudController@ajaxNames)

**Comment by tabacitu** [comment]
Guys, 

I've been working on filters and I'm really happy with the progress. The syntax is very similar to the above: very powerful and it's very easily customizable. The prototype is ready on a feature branch (only select and select2 right now) and I'm pretty sure you'll love it.

I am, however, struggling with one interface dilemma and I'd like a second opinion. To recap, there are 4 places ("stacks") where you can place filters: top, right, bottom, left. I expect the most used to be "top" and "right", because that's where it makes most sense.

Here's how they all look:
![screen shot 2016-11-08 at 18 57 23](https://cloud.githubusercontent.com/assets/1032474/20108830/2349ea48-a5e6-11e6-9569-7755c0298d19.png)
![screen shot 2016-11-08 at 18 57 39](https://cloud.githubusercontent.com/assets/1032474/20108832/234a2788-a5e6-11e6-95cb-d2b9aa4d45aa.png)
![screen shot 2016-11-08 at 18 57 51](https://cloud.githubusercontent.com/assets/1032474/20108831/2349f768-a5e6-11e6-9209-2e36828eacbc.png)
![screen shot 2016-11-08 at 18 57 05](https://cloud.githubusercontent.com/assets/1032474/20108829/2346c732-a5e6-11e6-8b80-c052f227e088.png)

I like how "right" and "left" look. But I'm not very happy with "top" and "bottom". I've also tried it this way, it looks cleaner, but it seems they take up too much space, if the label and select are on separate rows:
![screen shot 2016-11-08 at 18 59 11](https://cloud.githubusercontent.com/assets/1032474/20108877/5f5e3dae-a5e6-11e6-9fa0-a2bc937cfc61.png)

I've also thought about having them in a drawer menu (hidden, click "Filters" and a drawer opens up from the right), but it's a two-click process instead of a one-click process and that might be annoying. Plus, it opens up the issue of letting the user know somehow that some filters are applied.

Any ideas/suggestions? Do you think it's ok to leave "top" like that?

Thanks a lot. Cheers!

**Comment by properius** [comment]
I like AngelList's top filters for their jobs search ([https://angel.co/jobs](https://angel.co/jobs)):

![filters](https://cloud.githubusercontent.com/assets/6814904/20110684/a1a2ee72-a5e5-11e6-8b86-ae1b2ed52e8c.png)

User always knows which filters are applied. Even with a lot of filters selected they don't take up too much space.

There is also an option to save filters. Most filters have option to include or exclude values.

Drop-downs for filters are optimized based on their values:

![filter1](https://cloud.githubusercontent.com/assets/6814904/20110864/5df11428-a5e6-11e6-86d9-db2b45cf8148.png)
--
![filter2](https://cloud.githubusercontent.com/assets/6814904/20110865/5df153e8-a5e6-11e6-95eb-f6cc5e285995.png)
--
![filter3](https://cloud.githubusercontent.com/assets/6814904/20110867/5df4d04a-a5e6-11e6-8570-8cf0f59509ae.png)
--
![filter4](https://cloud.githubusercontent.com/assets/6814904/20110866/5df34806-a5e6-11e6-8282-b5611ccb4055.png)
--
![filter5](https://cloud.githubusercontent.com/assets/6814904/20110863/5df09e6c-a5e6-11e6-83dc-814910a5bacb.png)
--

Just an idea since you asked :-)  Your design works too. Maybe for top/bottom the filters could be included in the same box as the table. Maybe below the top buttons.

**Comment by tabacitu** [comment]
@properius ... you rule, man. I can't thank you enough.

I really _really_ like the AngelList filters, but I dismissed them at first. I thought it would make it far too difficult for developers to customize and make their own filters. But I played around for a few hours just to make sure and... I think I ended up with something excellent.

So the only thing that would make it very very difficult to implement in Backpack is the "tag" system - every filter gets added as a tag in that input. It would make it a nightmare to handle when building custom filters, which I imagine will happen a lot. Customization is a very important feature of Backpack. And it's also the one thing that makes the filtering a little bit confusing there: you activate the filter from one spot but then deactivate it from another spot.

BUT. I played around with different ways to show the "active filters" and I really think you'll like my solution. So without further ado, I present to you... (drums rolling): **the bootstrap navbar**.
![screen shot 2016-11-09 at 11 29 06](https://cloud.githubusercontent.com/assets/1032474/20133317/c47d6a3c-a66f-11e6-8742-b4200ae9684e.png)

I know, right?! It was literally in front of our eyes this whole time. I found it might be a perfect solution for our filters interface. I'm still testing its reliability, but right now I am very _very_ happy with this solution and I wanted to show you some screenshots. Click to open them full-size for a closer-to-reality experience.

![screen shot 2016-11-09 at 11 20 17](https://cloud.githubusercontent.com/assets/1032474/20134010/75f3af9a-a672-11e6-852f-685f85e452ae.png)
![screen shot 2016-11-09 at 11 20 29](https://cloud.githubusercontent.com/assets/1032474/20134011/75f45dbe-a672-11e6-98b1-bf6ab5f1d3c7.png)
![screen shot 2016-11-09 at 11 20 39](https://cloud.githubusercontent.com/assets/1032474/20134013/75f9bc1e-a672-11e6-9737-529cad34df21.png)
![screen shot 2016-11-09 at 11 20 54](https://cloud.githubusercontent.com/assets/1032474/20134012/75f799ca-a672-11e6-9a79-b1d112563454.png)
![screen shot 2016-11-09 at 11 21 05](https://cloud.githubusercontent.com/assets/1032474/20134014/75fac028-a672-11e6-8023-2a7b51bf3d21.png)


Here's what I think are the PROs and CONs of using the navbar vs using the filters above:

PROs | CONs
--- | --- 
prettier | -
collapsed on mobile by default | -
eliminates the need for multiple stacks (top, left, bottom, right) | -
they fit nicely in the perfect spot for filters: right above the datatable | -
are both very visible AND out-of-the-way | -
active filters are easy to spot using the "active" state | -
allows for very simple filters, using Eloquent scopes (Trashed, Published, etc) | -
allows for very complex filters with no added space, because the actual filter will be inside a dropdown (think x > y OR z >= t ) | -
will still look good with 20+ filters, on multiple rows | -
coding a custom dropdown (filter) should be very familiar to developers, since it's just a bootstrap navbar | coding a custom filter WILL have more complicated JS than before and will be pretty different from custom fields 
- | on the JS side, it will be kind of hacky to place values on ```<li>``` elements or ```<input type="hidden">``` inside these elements


As you might notice, I am very very biased towards switching to this new system. If you see something against it (or an even better solution), please speak now. It will be very very difficult to change it after we launch it. Also, feedback please (of course).

And @properius , thanks again :-)

Cheers!

**Comment by mosid** [comment]
You didn't left us something to say.
This is just awesome....

**Comment by tabacitu** [comment]
I'm done implementing the most important filters and will be launching the feature soon. Any feedback on the filter docs? https://laravel-backpack.readme.io/v3.0/docs/filters#section-filter-types

Cheers!

**Comment by tabacitu** [comment]
Done, launched! Use it and abuse it :-)

**Comment by zakirullin** [comment]
Thanks for filters feature! What about date-range filters?
